QoS - Expose the assert_liveliness API for Publishers and Nodes

emersonknapp opened the pull request:
Depends on ros2/rmw#171

Expose the rcl assert_liveliness APIs for Nodes and Publishers

Signed-off-by: Emerson Knapp <eknapp@amazon.com>

nburek commented:
@wjwwood This is ready for review. Can you please update the label? Thank you.

jacobperron commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -3629,6 +3629,57 @@ rclpy_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+
+  if (PyCapsule_IsValid(pyentity, "rcl_node_t")) {
+    rcl_node_t * node = (rcl_node_t *)PyCapsule_GetPointer(pyentity, "rcl_node_t");
+    if (!node) {
```
This guard is redundant and can be removed. `PyCapsule_IsValid` does the same checks as `PyCapsule_GetPointer`.

jacobperron commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -3629,6 +3629,57 @@ rclpy_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+  } else if (PyCapsule_IsValid(pyentity, "rcl_publisher_t")) {
+    rcl_publisher_t * publisher = (rcl_publisher_t *)PyCapsule_GetPointer(
+      pyentity, "rcl_publisher_t");
+    if (!publisher) {
```
Redundant check. See previous comment.

jacobperron commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -3629,6 +3629,57 @@ rclpy_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+    }
+    if (RCL_RET_OK != rcl_node_assert_liveliness(node)) {
+      PyErr_Format(PyExc_RuntimeError,
+        "Failed assert liveliness on the node.");
```
There is more than one reason why we could fail to assert liveliness (e.g. invalid node). Consider appending the `rcl` error:
```c
PyErr_Format(PyExc_RuntimeError,
        "Failed assert liveliness on the node: %s", rcl_get_error_string().str);
```

jacobperron commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -3629,6 +3629,57 @@ rclpy_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+    }
+    if (RCL_RET_OK != rcl_publisher_assert_liveliness(publisher)) {
+      PyErr_Format(PyExc_RuntimeError,
+        "Failed to assert liveliness on the publisher.");
```
Consider appending the `rcl` error string.

emersonknapp commented:
Thanks for the feedback! I've addressed the comments in a pushed commit. If we're happy with how this looks - we can leave it on deck for CI for when the bundle of dependencies ros2/rmw#171 is merged.

emersonknapp commented:
@thomas-moulard - please run the following CI job:

 * Gist file: https://gist.githubusercontent.com/emersonknapp/02889df53fbb6e7b06ebda807d930d15/raw/5a135ec8266374e6b9e7ae3a2e54b4d0a7fa67b9/ros2.repos
 * Additional BUILD flags: --packages-up-to rclpy
 * Additional TEST flags: --packages-select rclpy
* CI Job: ci_launcher

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6942)](http://ci.ros2.org/job/ci_linux/6942/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3203)](http://ci.ros2.org/job/ci_linux-aarch64/3203/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5699)](http://ci.ros2.org/job/ci_osx/5699/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6771)](http://ci.ros2.org/job/ci_windows/6771/)

emersonknapp commented:
@thomas-moulard I messed up and pointed that CI against the _other_ Python PR - I have updated the existing Gist to be correct - here it is. So sorry about that

https://gist.githubusercontent.com/emersonknapp/02889df53fbb6e7b06ebda807d930d15/raw/789129bd6b869b0bbc16a5f1f2b890da87e711c3/ros2.repos

tfoote commented:
Retriggered with the updated repos file:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6944)](http://ci.ros2.org/job/ci_linux/6944/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3204)](http://ci.ros2.org/job/ci_linux-aarch64/3204/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5701)](http://ci.ros2.org/job/ci_osx/5701/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6772)](http://ci.ros2.org/job/ci_windows/6772/)

sloretz commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -3604,6 +3604,53 @@ rclpy_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+      rcl_reset_error();
+      return NULL;
+    }
+  } else if (PyCapsule_IsValid(pyentity, "rcl_publisher_t")) {
```
`rcl_publisher_t` instead of `rclpy_publisher_t` since #319. Mind rebasing this PR?

https://github.com/ros2/rclpy/blob/63cbc5fdef23ea86ec336e8d886c028238ce54f6/rclpy/src/rclpy/_rclpy.c#L50-L54

emersonknapp replied:
Done

sloretz commented:
CI (Testing just rclpy)
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6955)](http://ci.ros2.org/job/ci_linux/6955/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3208)](http://ci.ros2.org/job/ci_linux-aarch64/3208/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5706)](http://ci.ros2.org/job/ci_osx/5706/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6779)](http://ci.ros2.org/job/ci_windows/6779/)

emersonknapp commented:
@sloretz hurray!

sloretz commented on rclpy/src/rclpy/_rclpy.c:
```diff
@@ -3604,6 +3604,53 @@ rclpy_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+      return NULL;
+    }
+  } else if (PyCapsule_IsValid(pyentity, "rclpy_publisher_t")) {
+    rcl_publisher_t * publisher = (rcl_publisher_t *)PyCapsule_GetPointer(
```
This works because `rcl_publisher_t` is the first member of `rclpy_publisher_t`, but it would be nice to be more explicit about it

emersonknapp replied:
I don't disagree - do we want to block merging for that? Or can we follow up afterwards?

sloretz replied:
I'll merge this one.

sloretz commented:
Thanks for the PR!
